cmd/image-builder/main.go

```diff
@@ -162,15 +185,163 @@ func runInKaniko(o options, name string, destinations, platforms []string, build
+	if !present {
+		return nil, fmt.Errorf("PULL_BASE_SHA environment variable is not set, please set it to valid pull base SHA")
+	}
+	templateParameters.SetBaseSHA(baseSHA)
+
+	pullSHA, present := os.LookupEnv("PULL_PULL_SHA")
+	if present {
+		templateParameters.SetPullSHA(pullSHA)
+	}
+
+	templateParameters.SetImageName(options.Name)
+
+	templateParameters.SetDockerfilePath(options.Dockerfile)
+
+	templateParameters.SetBuildContext(options.Context)
+
+	templateParameters.SetExportTags(options.ExportTags)
+
+	if len(options.BuildArgs) > 0 {
+		templateParameters.SetBuildArgs(options.BuildArgs.String())
+	}
+
+	if len(options.Tags) > 0 {
+		templateParameters.SetImageTags(options.Tags.String())
+	}
+
+	err := templateParameters.Validate()
+	if err != nil {
+		return nil, fmt.Errorf("failed validating ADO template parameters, err: %w", err)
+	}
+
+	return templateParameters, nil
+}
+
+// TODO(dekiel): refactor this function to accept clients as parameters to make it testable with mocks.
+func buildInADO(o options) error {
+	fmt.Println("Building image in ADO pipeline.")
+	adoPAT, present := os.LookupEnv("ADO_PAT")
+	if !present {
+		return fmt.Errorf("build in ADO failed, ADO_PAT environment variable is not set, please set it to valid ADO PAT")
+	}
+
+	fmt.Println("Preparing ADO template parameters.")
+	templateParameters, err := prepareADOTemplateParameters(o)
+	if err != nil {
+		return fmt.Errorf("build in ADO failed, failed preparing ADO template parameters, err: %s", err)
+	}
+
+	fmt.Println("Running ADO pipeline.")
+	adoClient := adopipelines.NewClient(o.ADOOrganizationURL, adoPAT)
+
+	fmt.Println("Triggering ADO build pipeline")
+	ctx := context.Background()
+	pipelineRun, err := adopipelines.Run(ctx, adoClient, templateParameters, o.GetADOConfig())
+	if err != nil {
+		return fmt.Errorf("build in ADO failed, failed running ADO pipeline, err: %s", err)
+	}
+
+	pipelineRunResult, err := adopipelines.GetRunResult(ctx, adoClient, o.GetADOConfig(), pipelineRun.Id, 30*time.Second)
+	if err != nil {
+		return fmt.Errorf("build in ADO failed, failed getting ADO pipeline run result, err: %s", err)
+	}
+	fmt.Printf("ADO pipeline run finished with status: %s", *pipelineRunResult)
+
+	fmt.Println("Getting ADO pipeline run logs.")
+	adoBuildClient, err := adopipelines.NewBuildClient(o.ADOOrganizationURL, adoPAT)
+	if err != nil {
+		fmt.Printf("Can't read ADO pipeline run logs, failed creating ADO build client, err: %s", err)
+	}
+	logs, err := adopipelines.GetRunLogs(ctx, adoBuildClient, &http.Client{}, o.GetADOConfig(), pipelineRun.Id, adoPAT)
+	if err != nil {
+		fmt.Printf("Failed read ADO pipeline run logs, err: %s", err)
+	} else {
+		fmt.Printf("ADO pipeline image build logs:\n%s", logs)
+	}
+
+	if *pipelineRunResult == pipelines.RunResultValues.Failed || *pipelineRunResult == pipelines.RunResultValues.Unknown {
+		return fmt.Errorf("build in ADO finished with status: %s", *pipelineRunResult)
+	}
+	return nil
+}
```

Comment:
In my opinion this kind of specific configuration should be handled in separate typedef. Then you parse the template path to ado template parsed separately. This IMO should only be parsed on demand only when you use ADO-based configuration.
See runFunc and the definition this function implements. You pass `options` here, and you can run parsing of a template inside this designated function.

Reply:
I don't fully understand this comment.
This function is executed only if user run a tool with build-in-ado flag enabled, so only when requested.
There is no need to parse any template.
This part is collecting values which are needed to trigger oci-image-builder ado pipeline. After collecting this values a ado api is called to trigger a build. That's it.